diff --git a/cs.c b/cs.c
index c6fa5f9..95cce96 100644
--- a/cs.c
+++ b/cs.c
@@ -707,10 +707,12 @@ cs_err CAPSTONE_API cs_option(csh ud, cs_opt_type type, size_t value)
 			return CS_ERR_OK;
 
 		case CS_OPT_SKIPDATA_SETUP:
-			if (value)
+			if (value) {
 				handle->skipdata_setup = *((cs_opt_skipdata *)value);
-				if (handle->skipdata_setup.mnemonic == NULL)
+				if (handle->skipdata_setup.mnemonic == NULL) {
 					handle->skipdata_setup.mnemonic = SKIPDATA_MNEM;
+				}
+			}
 			return CS_ERR_OK;
 
 		case CS_OPT_MNEMONIC:
