diff --git a/arch/AArch64/AArch64GenAsmWriter.inc b/arch/AArch64/AArch64GenAsmWriter.inc
index 0cfe199..de9c4d0 100644
--- a/arch/AArch64/AArch64GenAsmWriter.inc
+++ b/arch/AArch64/AArch64GenAsmWriter.inc
@@ -26297,7 +26296,7 @@ static char *printAliasInstr(MCInst *MI, SStream *OS, MCRegisterInfo *MRI)
       }
     } while (AsmString[I] != '\0');
   }
-  free(AsmString);
+  cs_mem_free(AsmString);
   return tmpString;
 }
         
diff --git a/suite/synctools/asmwriter.py b/suite/synctools/asmwriter.py
index 9636c13..b1d18b5 100755
--- a/suite/synctools/asmwriter.py
+++ b/suite/synctools/asmwriter.py
@@ -788,7 +788,7 @@ for line in lines:
       }
     } while (AsmString[I] != '\\0');
   }
-  free(AsmString);
+  cs_mem_free(AsmString);
   return tmpString;
 }
         """)
diff --git a/utils.c b/utils.c
index 194e070..efb9da7 100644
--- a/utils.c
+++ b/utils.c
@@ -91,7 +91,7 @@ unsigned int count_positive8(const unsigned char *list)
 
 char *cs_strdup(const char *str)
 {
-	size_t len = strlen(str)+ 1;
+	size_t len = strlen(str) + 1;
 	void *new = cs_mem_malloc(len);
 
 	if (new == NULL)
