diff --git a/include/capstone/arm.h b/include/capstone/arm.h
new file mode 100644
index 00000000..5445c96e
--- /dev/null
+++ b/include/capstone/arm.h
@@ -0,0 +1 @@
+#include "../arm.h"
diff --git a/include/capstone/arm64.h b/include/capstone/arm64.h
new file mode 100644
index 00000000..30b49da7
--- /dev/null
+++ b/include/capstone/arm64.h
@@ -0,0 +1 @@
+#include "../arm64.h"
diff --git a/include/capstone/capstone.h b/include/capstone/capstone.h
new file mode 100644
index 00000000..761a67b8
--- /dev/null
+++ b/include/capstone/capstone.h
@@ -0,0 +1 @@
+#include "../capstone.h"
diff --git a/include/capstone/mips.h b/include/capstone/mips.h
new file mode 100644
index 00000000..b91dbabb
--- /dev/null
+++ b/include/capstone/mips.h
@@ -0,0 +1 @@
+#include "../mips.h"
diff --git a/include/capstone/platform.h b/include/capstone/platform.h
new file mode 100644
index 00000000..40f1b399
--- /dev/null
+++ b/include/capstone/platform.h
@@ -0,0 +1 @@
+#include "../platform.h"
diff --git a/include/capstone/ppc.h b/include/capstone/ppc.h
new file mode 100644
index 00000000..d18092c9
--- /dev/null
+++ b/include/capstone/ppc.h
@@ -0,0 +1 @@
+#include "../ppc.h"
diff --git a/include/capstone/sparc.h b/include/capstone/sparc.h
new file mode 100644
index 00000000..e6b9d11b
--- /dev/null
+++ b/include/capstone/sparc.h
@@ -0,0 +1 @@
+#include "../sparc.h"
diff --git a/include/capstone/systemz.h b/include/capstone/systemz.h
new file mode 100644
index 00000000..96160def
--- /dev/null
+++ b/include/capstone/systemz.h
@@ -0,0 +1 @@
+#include "../systemz.h"
diff --git a/include/capstone/x86.h b/include/capstone/x86.h
new file mode 100644
index 00000000..1f504270
--- /dev/null
+++ b/include/capstone/x86.h
@@ -0,0 +1 @@
+#include "../x86.h"
diff --git a/include/capstone/xcore.h b/include/capstone/xcore.h
new file mode 100644
index 00000000..1482ce50
--- /dev/null
+++ b/include/capstone/xcore.h
@@ -0,0 +1 @@
+#include "../xcore.h"
