diff --git a/shlr/qjs/src/quickjs.c b/shlr/qjs/src/quickjs.c
index ddb076e0c7..c3714b44d2 100644
--- a/shlr/qjs/src/quickjs.c
+++ b/shlr/qjs/src/quickjs.c
@@ -42511,7 +42511,7 @@ static JSValue js___date_clock(JSContext *ctx, JSValueConst this_val,
 /* OS dependent. d = argv[0] is in ms from 1970. Return the difference
    between UTC time and local time 'd' in minutes */
 static int getTimezoneOffset(int64_t time) {
-#if defined(_WIN32)
+#if defined(_WIN32) || defined(_AIX)
     /* XXX: TODO */
     return 0;
 #else
diff --git a/shlr/qjs/src/quickjs.h b/shlr/qjs/src/quickjs.h
index 9e55cb837c..60f559c790 100644
--- a/shlr/qjs/src/quickjs.h
+++ b/shlr/qjs/src/quickjs.h
@@ -73,6 +73,14 @@ typedef struct JSRefCountHeader {
     int ref_count;
 } JSRefCountHeader;
 
+#if defined(_AIX) && defined(__GNUC__)
+/* Make NAN and INFINITY constant expressions */
+#undef NAN
+#define NAN __builtin_nanf("0x7fc00000")
+#undef INFINITY
+#define INFINITY __builtin_huge_valf()
+#endif
+
 #define JS_FLOAT64_NAN NAN
 
 #if defined(JS_STRICT_NAN_BOXING)
