diff --git a/arch/X86/X86Disassembler.c b/arch/X86/X86Disassembler.c
index ecf8142..76523d9 100644
--- a/arch/X86/X86Disassembler.c
+++ b/arch/X86/X86Disassembler.c
@@ -911,17 +911,7 @@ bool X86_getInstruction(csh ud, const uint8_t *code, size_t code_len,
 	info.offset = address;
 
 	memset(&insn, 0, offsetof(InternalInstruction, reader));
-	//initialize some the necessary values
-	insn.prefixPresent[0x26] = 0;
-	insn.prefixPresent[0x2e] = 0;
-	insn.prefixPresent[0x36] = 0;
-	insn.prefixPresent[0x3e] = 0;
-	insn.prefixPresent[0x64] = 0;
-	insn.prefixPresent[0x65] = 0;
-	insn.prefixPresent[0x66] = 0;
-	insn.prefixPresent[0xf0] = 0;
-	insn.prefixPresent[0xf2] = 0;
-	insn.prefixPresent[0xf3] = 0;
+	memset(&insn.prefixPresent, 0, sizeof(insn.prefixPresent));
 
 	if (instr->flat_insn->detail) {
 		instr->flat_insn->detail->x86.op_count = 0;
