diff -r 5dd73dc3dbf1 swig/vapi/r_bin.vapi
--- a/swig/vapi/r_bin.vapi	Thu Feb 17 16:14:27 2011 +0100
+++ b/swig/vapi/r_bin.vapi	Thu Feb 17 21:28:36 2011 +0100
@@ -42,7 +42,8 @@
 		public int has_dbg_linenums();
 		public int has_dbg_syms();
 		public int has_dbg_relocs();
-		public int meta_get_line(uint64 addr, ref string file, int len, out int line);
+		public int meta_get_line(uint64 addr, out string file, int len, out int line);
+		//XXX: workaround for GUILE public int meta_get_line(uint64 addr, ref string file, int len, out int line);
 		public string meta_get_source_line(uint64 addr);
 	
 		[CCode (cname="RBinArch", free_function="", ref_function="", unref_function="")]
diff -r 5dd73dc3dbf1 swig/vapi/r_parse.vapi
--- a/swig/vapi/r_parse.vapi	Thu Feb 17 16:14:27 2011 +0100
+++ b/swig/vapi/r_parse.vapi	Thu Feb 17 21:28:36 2011 +0100
@@ -9,9 +9,15 @@
 
 		public int list();
 		public bool use(string name);
+/* XXX this must be ref... but SWIG fails with the GUILE backend .. workaround */
+		public bool filter(RFlag flag, out string data, out string str, int len);
+		public bool assemble(out string dst, out string src);
+		public bool parse(void *dst, out string src);
+/*
 		public bool filter(RFlag flag, ref string data, ref string str, int len);
 		public bool assemble(ref string dst, ref string src);
 		public bool parse(void *dst, ref string src);
+*/
 /*
 
 		public void set_user_ptr(void *user);
